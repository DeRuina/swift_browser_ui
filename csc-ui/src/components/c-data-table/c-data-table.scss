:host {
  display: block;
  position: relative;
  width: 100%;
}

:host([horizontal-scrolling]) {
  overflow-x: scroll;
}

table,
caption,
tbody,
tfoot,
thead,
tr,
th,
td {
  border: 0;
  font-size: 100%;
  font: inherit;
  margin: 0;
  padding: 0;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

.c-data-table {
  box-shadow: inset 0 0 0 1px var(--csc-light-grey);
  width: 100%;

  .children {
    display: flex;
    flex-wrap: wrap;
    gap: 4px;
    width: 100%;

    pre {
      font-family: inherit;
      white-space: pre-wrap;
    }
  }

  [data-justify='end'],
  [data-align='end'] .children {
    align-items: flex-end;
  }

  [data-justify='start'],
  [data-align='start'] .children {
    align-items: flex-start;
  }

  [data-justify='center'],
  [data-align='center'] .children {
    align-items: center;
  }

  [data-align='end'],
  [data-justify='end'] .children {
    justify-content: flex-end;
  }

  [data-align='start'],
  [data-justify='start'] .children {
    justify-content: flex-start;
  }

  [data-align='center'],
  [data-justify='center'] .children {
    justify-content: center;
  }

  th,
  td {
    font-weight: 400;

    > div {
      align-items: center;
      box-sizing: border-box;
      display: flex;
      line-height: normal;
      padding: 12px;
      width: 100%;
    }
  }

  th {
    font-size: 14px;
    padding: 4px;

    &.selection {
      width: 56px;
    }

    &.indicator {
      width: 46px;
    }

    > div {
      border-radius: 4px;
      flex-wrap: nowrap;
      gap: 8px;
      height: 48px;
      white-space: nowrap;

      &.sortable:hover {
        background-color: var(--csc-primary-text-hover);
        color: var(--csc-primary);
        cursor: pointer;

        svg {
          fill: var(--csc-primary);
        }
      }

      &.selection {
        &--heading {
          padding-left: 8px;
          padding-right: 0;
          margin-right: -8px;
        }
      }
    }
  }

  td {
    font-size: 16px;

    > div {
      align-items: flex-start;
      flex-direction: column;
      gap: 8px;
      justify-content: center;
      min-height: 56px;
      padding: 12px;

      &.selection {
        margin-right: -4px;
        padding: 7px 0 7px 12px;
      }

      c-tag {
        --c-tag-margin: 0px;
      }
    }

    &.loader {
      background-color: var(--csc-light-grey);
      height: 2px;
      position: relative;

      > div {
        height: 2px;
        min-height: auto;
        padding: 0;
      }
    }

    &.info {
      > div {
        height: 108px;
        justify-content: center;
      }
    }

    svg {
      transition: transform 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);
    }
  }

  thead {
    background-color: #fff;
    position: relative;

    &.sticky {
      box-shadow: 0 2px 0 0 var(--csc-light-grey);
      position: sticky;
      top: 0;
      z-index: 3;
    }

    .c-data-table__header-indicators {
      pointer-events: none;

      th {
        > div {
          margin-top: -56px;
          display: flex;
          justify-items: space-between;
        }
      }

      c-icon-button {
        background-color: rgba(255, 255, 255, 0.75);
        border-radius: 50%;
        pointer-events: all;
      }
    }
  }

  tfoot {
    background-color: #fff;

    tr {
      box-shadow: inset 0 1px 0 0 var(--csc-light-grey);
    }

    td {
      > div {
        align-items: center;
        flex-direction: row;
        justify-content: flex-start;
        min-height: 48px;
      }

      c-pagination {
        flex: 1;
      }
    }
  }

  tbody {
    /* Safari box shadow replacement */
    @supports (-webkit-hyphens: none) {
      tr {
        border-bottom: 1px solid var(--csc-light-grey);
      }
    }

    tr {
      box-shadow: inset 0 1px 0 0 var(--csc-light-grey);

      &.actions {
        box-shadow: none;

        td > div {
          background-color: rgba(0, 0, 0, 0.015);
          height: auto;
        }
      }

      &.active {
        box-shadow: inset 2px 0 0 var(--csc-primary),
          inset -2px 0 0 var(--csc-primary), inset 0 2px 0 var(--csc-primary);
        border-radius: 6px 6px 0 0;

        svg {
          transform: rotate(180deg);
        }

        + .additional-data {
          border-radius: 0 0 6px 6px;
          box-shadow: inset 2px 0 0 var(--csc-primary),
            inset -2px 0 0 var(--csc-primary), inset 0 -2px 0 var(--csc-primary);
        }

        &.selected {
          box-shadow: inset 4px 0 0 var(--csc-primary),
            inset -2px 0 0 var(--csc-primary), inset 0 2px 0 var(--csc-primary);
        }
      }

      &.selected {
        background-color: rgba($csc-primary, 0.05);
        box-shadow: inset 4px 0 0 var(--csc-primary),
          inset 0 1px 0 0 var(--csc-light-grey);

        + .additional-data {
          box-shadow: inset 4px 0 0 var(--csc-primary),
            inset -2px 0 0 var(--csc-primary), inset 0 -2px 0 var(--csc-primary);
        }
      }
    }
  }

  &__loader {
    background-color: var(--csc-light-grey);
    height: 2px;
    overflow: hidden;
    position: absolute;
    width: 100%;
  }

  .loading-bar {
    animation: loading 1s ease-in 0.5s infinite;
    background-color: var(--csc-primary);
    height: 100%;
    left: -50%;
    position: absolute;
    width: 50%;
  }

  @keyframes loading {
    0% {
      transform: translateX(0);
    }
    to {
      transform: translateX(400%);
    }
  }

  .additional-data {
    display: none;

    &.active {
      display: table-row;
    }

    ul {
      list-style: none;
      margin: 0;
      padding: 0;
      width: 100%;

      li {
        padding: 12px;

        &:not(:last-child) {
          box-shadow: inset 0 -1px 0 0 var(--csc-light-grey);
        }

        span {
          color: var(--csc-primary);
          font-weight: 600;
        }

        .title {
          display: flex;
          gap: 8px;

          + .children {
            margin-top: 8px;
          }
        }
      }
    }
  }

  &--scrollable {
    position: relative;
  }

  &--hoverable {
    tbody {
      tr:not(.additional-data) {
        cursor: pointer;

        &:hover {
          background-color: var(--csc-primary-text-hover);
          border-radius: 6px;
          box-shadow: inset 0 0 0 2px var(--csc-primary);
          overflow: hidden;
        }

        &.active:hover {
          border-radius: 6px 6px 0 0;
          box-shadow: inset 2px 0 0 var(--csc-primary),
            inset -2px 0 0 var(--csc-primary), inset 0 2px 0 var(--csc-primary);
        }

        &.selected.active:hover {
          border-radius: 6px 6px 0 0;
          box-shadow: inset 4px 0 0 var(--csc-primary),
            inset -2px 0 0 var(--csc-primary), inset 0 2px 0 var(--csc-primary);
        }

        &.selected:hover {
          border-radius: 6px;
          box-shadow: inset 4px 0 0 var(--csc-primary),
            inset 0 0 0 2px var(--csc-primary);
        }
      }
    }
  }
}

.sorted-column {
  background-color: rgb(0 103 120 / 5%);
}
