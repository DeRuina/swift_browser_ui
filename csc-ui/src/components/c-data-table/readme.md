# c-datatable



<!-- Auto Generated Below -->


## Properties

| Property              | Attribute              | Description                                               | Type                      | Default                                                                                                                                  |
| --------------------- | ---------------------- | --------------------------------------------------------- | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| `data`                | --                     | Data of the table                                         | `CDataTableData[]`        | `[]`                                                                                                                                     |
| `externalData`        | `external-data`        | Externally sorted and paginated data                      | `boolean`                 | `false`                                                                                                                                  |
| `footerOptions`       | --                     | Items per page options                                    | `CDataTableFooterOptions` | `{     itemsPerPageOptions: [5, 25, 50, 100],     hideDetails: false,     simple: false,     hideRange: false,     size: 'default',   }` |
| `headers`             | --                     | Headers of the table                                      | `CDataTableHeader[]`      | `[]`                                                                                                                                     |
| `hideFooter`          | `hide-footer`          | Hide the footer                                           | `boolean`                 | `false`                                                                                                                                  |
| `horizontalScrolling` | `horizontal-scrolling` | Use horizontal scrolling                                  | `boolean`                 | `false`                                                                                                                                  |
| `loading`             | `loading`              | Show a loader on top of the table                         | `boolean`                 | `false`                                                                                                                                  |
| `loadingText`         | `loading-text`         | Text shown when there is no data and the table is loading | `string`                  | `'Loading data'`                                                                                                                         |
| `noDataText`          | `no-data-text`         | Text shown when there are no data available               | `string`                  | `'No data'`                                                                                                                              |
| `pagination`          | --                     | Pagination options                                        | `CPaginationOptions`      | `undefined`                                                                                                                              |
| `selectable`          | `selectable`           | Make rows selectable                                      | `boolean`                 | `false`                                                                                                                                  |
| `selectionProperty`   | `selection-property`   | Property used in selections                               | `string`                  | `null`                                                                                                                                   |
| `singleExpansion`     | `single-expansion`     | Allow only a single row expanded at a time                | `boolean`                 | `false`                                                                                                                                  |
| `singleSelection`     | `single-selection`     | Select only a single row at a time                        | `boolean`                 | `false`                                                                                                                                  |
| `sortBy`              | `sort-by`              | Sort data by                                              | `any`                     | `null`                                                                                                                                   |
| `sortDirection`       | `sort-direction`       | Sorting direction                                         | `"asc" \| "desc"`         | `null`                                                                                                                                   |
| `stickyHeader`        | `sticky-header`        | Use sticky header                                         | `boolean`                 | `false`                                                                                                                                  |


## Events

| Event       | Description                | Type                              |
| ----------- | -------------------------- | --------------------------------- |
| `expand`    | Triggered on row expansion | `CustomEvent<any>`                |
| `paginate`  | Triggered on pagination    | `CustomEvent<CPaginationOptions>` |
| `selection` | Triggered on selection     | `CustomEvent<any>`                |
| `sort`      | Triggered on sort          | `CustomEvent<any>`                |


## Methods

### `clearSelections() => Promise<void>`

Clear selections externally

#### Returns

Type: `Promise<void>`



### `getData() => Promise<Record<string, string | number | boolean>[]>`

Provide sorted data

#### Returns

Type: `Promise<Record<string, string | number | boolean>[]>`




## Dependencies

### Depends on

- [c-checkbox](../c-checkbox)
- [c-pagination](../c-pagination)
- [c-icon-button](../c-icon-button)
- [c-icon](../c-icon)
- [c-spacer](../c-spacer)

### Graph
```mermaid
graph TD;
  c-data-table --> c-checkbox
  c-data-table --> c-pagination
  c-data-table --> c-icon-button
  c-data-table --> c-icon
  c-data-table --> c-spacer
  c-pagination --> c-menu
  c-pagination --> c-icon-button
  c-menu --> c-menu-items
  style c-data-table fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
