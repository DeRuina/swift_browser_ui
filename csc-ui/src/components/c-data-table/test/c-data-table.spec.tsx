import { newSpecPage } from '@stencil/core/testing';
import { CDataTable } from '../c-data-table';
import { basicData, basicHeaders } from './data';
import { h } from '@stencil/core';

declare global {
  interface Window {
    IntersectionObserver: unknown;
  }
}

describe('c-data-table', () => {
  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    global.IntersectionObserver = class IntersectionObserver {
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      constructor() {}

      disconnect() {
        return null;
      }

      observe() {
        return null;
      }

      takeRecords() {
        return null;
      }

      unobserve() {
        return null;
      }
    };
    global.ResizeObserver = class ResizeObserver {
      disconnect() {
        return null;
      }

      observe() {
        return null;
      }

      takeRecords() {
        return null;
      }

      unobserve() {
        return null;
      }
    };
  });
  it.only('renders default html', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      template: () => (
        <c-data-table headers={basicHeaders} data={basicData}></c-data-table>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });

  it.only('renders specific columns as hidden and expands on click', async () => {
    const headers = JSON.parse(JSON.stringify(basicHeaders));

    headers[2].hidden = true;

    const page = await newSpecPage({
      components: [CDataTable],
      template: () => (
        <c-data-table headers={headers} data={basicData}></c-data-table>
      ),
    });

    expect(page.root).toMatchSnapshot();

    const firstRow = page.root.shadowRoot.querySelector(
      '.parent',
    ) as HTMLElement;

    firstRow.click();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('renders horizontal scroll', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table ></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders disabled sorting', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('allows sorting by column', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders custom header width', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders pagination and adds custom pagination text', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders sticky header', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('allows to select rows', async () => {
    const page = await newSpecPage({
      components: [CDataTable],
      html: `<c-data-table></c-data-table>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
