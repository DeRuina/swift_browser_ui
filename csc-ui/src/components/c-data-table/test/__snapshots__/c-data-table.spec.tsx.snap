// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-data-table renders default html 1`] = `
<c-data-table>
  <mock:shadow-root>
    <table class="c-data-table">
      <colgroup>
        <col class="sorted-column">
        <col>
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th data-id="country">
            <div class="sortable">
              Country
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M7.03 9.97H11.03V18.89L13.04 18.92V9.97H17.03L12.03 4.97Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="population">
            <div class="sortable">
              Population (2020)
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M9,3L5,7H8V14H10V7H13M16,17V10H14V17H11L15,21L19,17H16Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="unemployment">
            <div class="sortable">
              Unemployment (%)
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M9,3L5,7H8V14H10V7H13M16,17V10H14V17H11L15,21L19,17H16Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="actions">
            <div></div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td class="loader" colspan="100"></td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              <div class="children">
                <p style="margin: 0;">
                  Denmark
                </p>
                <c-icon path="M14.4,6L14,4H5V21H7V14H12.6L13,16H20V6H14.4Z"></c-icon>
              </div>
            </div>
          </td>
          <td>
            <div>
              5831404
            </div>
          </td>
          <td>
            <div>
              4.8
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              Finland
            </div>
          </td>
          <td>
            <div>
              5529543
            </div>
          </td>
          <td>
            <div>
              7.5
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              Iceland
            </div>
          </td>
          <td>
            <div>
              366463
            </div>
          </td>
          <td>
            <div>
              5.4
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              Kermany
            </div>
          </td>
          <td>
            <div>
              18353442
            </div>
          </td>
          <td>
            <div>
              100.1
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              Norway
            </div>
          </td>
          <td>
            <div>
              5379475
            </div>
          </td>
          <td>
            <div>
              5
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div data-align="center">
              Sweden
            </div>
          </td>
          <td>
            <div>
              10353442
            </div>
          </td>
          <td>
            <div>
              8.7
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td colspan="100">
            <div class="c-data-table__footer">
              <c-pagination size="default"></c-pagination>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </mock:shadow-root>
</c-data-table>
`;

exports[`c-data-table renders specific columns as hidden and expands on click 1`] = `
<c-data-table>
  <mock:shadow-root>
    <table class="c-data-table c-data-table--hoverable">
      <colgroup>
        <col>
        <col class="sorted-column">
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th class="indicator"></th>
          <th data-id="country">
            <div class="sortable">
              Country
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M7.03 9.97H11.03V18.89L13.04 18.92V9.97H17.03L12.03 4.97Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="population">
            <div class="sortable">
              Population (2020)
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M9,3L5,7H8V14H10V7H13M16,17V10H14V17H11L15,21L19,17H16Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="actions">
            <div></div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td class="loader" colspan="100"></td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              <div class="children">
                <p style="margin: 0;">
                  Denmark
                </p>
                <c-icon path="M14.4,6L14,4H5V21H7V14H12.6L13,16H20V6H14.4Z"></c-icon>
              </div>
            </div>
          </td>
          <td>
            <div>
              5831404
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    4.8
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Finland
            </div>
          </td>
          <td>
            <div>
              5529543
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    7.5
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Iceland
            </div>
          </td>
          <td>
            <div>
              366463
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    5.4
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Kermany
            </div>
          </td>
          <td>
            <div>
              18353442
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    100.1
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Norway
            </div>
          </td>
          <td>
            <div>
              5379475
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    5
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Sweden
            </div>
          </td>
          <td>
            <div>
              10353442
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    8.7
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td colspan="100">
            <div class="c-data-table__footer">
              <c-pagination size="default"></c-pagination>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </mock:shadow-root>
</c-data-table>
`;

exports[`c-data-table renders specific columns as hidden and expands on click 2`] = `
<c-data-table>
  <mock:shadow-root>
    <table class="c-data-table c-data-table--hoverable">
      <colgroup>
        <col>
        <col class="sorted-column">
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th class="indicator"></th>
          <th data-id="country">
            <div class="sortable">
              Country
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M7.03 9.97H11.03V18.89L13.04 18.92V9.97H17.03L12.03 4.97Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="population">
            <div class="sortable">
              Population (2020)
              <svg height="24" tabindex="0" viewBox="0 0 24 24" width="24">
                <path d="M9,3L5,7H8V14H10V7H13M16,17V10H14V17H11L15,21L19,17H16Z"></path>
              </svg>
            </div>
          </th>
          <th data-id="actions">
            <div></div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td class="loader" colspan="100"></td>
        </tr>
        <tr class="active parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              <div class="children">
                <p style="margin: 0;">
                  Denmark
                </p>
                <c-icon path="M14.4,6L14,4H5V21H7V14H12.6L13,16H20V6H14.4Z"></c-icon>
              </div>
            </div>
          </td>
          <td>
            <div>
              5831404
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="active additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    4.8
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Finland
            </div>
          </td>
          <td>
            <div>
              5529543
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    7.5
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Iceland
            </div>
          </td>
          <td>
            <div>
              366463
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    5.4
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Kermany
            </div>
          </td>
          <td>
            <div>
              18353442
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    100.1
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Norway
            </div>
          </td>
          <td>
            <div>
              5379475
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    5
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
        <tr class="parent">
          <td>
            <div tabindex="0">
              <svg height="22" viewBox="0 0 24 24" width="22">
                <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
              </svg>
            </div>
          </td>
          <td>
            <div data-align="center">
              Sweden
            </div>
          </td>
          <td>
            <div>
              10353442
            </div>
          </td>
          <td>
            <div data-align="end">
              <div class="children">
                <c-button size="small" text="" title="Show Code">
                  Show Code
                </c-button>
              </div>
            </div>
          </td>
        </tr>
        <tr class="additional-data">
          <td colspan="100">
            <div>
              <ul>
                <li>
                  <div class="title">
                    <span>
                      Unemployment (%):
                    </span>
                    8.7
                  </div>
                </li>
              </ul>
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td colspan="100">
            <div class="c-data-table__footer">
              <c-pagination size="default"></c-pagination>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </mock:shadow-root>
</c-data-table>
`;
